fix(minifier): rotate associative operators to make it more idempotent

camc314 opened the pull request:
another **massive** pain to debug

~~https://github.com/oxc-project/monitor-oxc/actions/runs/12717213600~~

https://github.com/oxc-project/monitor-oxc/actions/runs/12717978983/job/35455601146

camc314 commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8424?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#8424** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8424?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8424?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#8420** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8420?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=8424).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

camc314 commented on crates/oxc_minifier/src/ast_passes/peephole_substitute_alternate_syntax.rs:
```diff
@@ -503,6 +504,29 @@ impl<'a, 'b> PeepholeSubstituteAlternateSyntax {
     }
 
+    /// `a || (b || c);` -> `(a || b) || c;`
+    fn try_rotate_logical_expression(
+        expr: &mut LogicalExpression<'a>,
+        ctx: Ctx<'a, 'b>,
+    ) -> Option<Expression<'a>> {
+        let Expression::LogicalExpression(right) = &mut expr.right else { return None };
+        if right.operator != expr.operator {
+            return None;
+        }
+        let new_expr = ctx.ast.expression_logical(
+            expr.span,
+            ctx.ast.expression_logical(
+                expr.span,
+                ctx.ast.move_expression(&mut expr.left),
+                expr.operator,
+                ctx.ast.move_expression(&mut right.left),
+            ),
+            expr.operator,
+            ctx.ast.move_expression(&mut right.right),
+        );
+        Some(new_expr)
+    }
```
causes an infinite loop somewhere 😞

camc314 replied:
technically not an infinite loop, this function needs to bulk change rather than doing 1x change per pass

camc314 replied:
😅 minification is hard. this should be fixed

camc314 commented:
@Boshen, minifier should now be impotent 🎉 

something is being mangled when it shouldn't though. trying to narrow it down now

 https://github.com/oxc-project/monitor-oxc/actions/runs/12717978983/job/35455601146

camc314 commented on crates/oxc_minifier/src/ast_passes/peephole_substitute_alternate_syntax.rs:
```diff
@@ -503,6 +504,38 @@ impl<'a, 'b> PeepholeSubstituteAlternateSyntax {
     }
 
+    /// `a || (b || c);` -> `(a || b) || c;`
+    fn try_rotate_logical_expression(
```
TODO: need to improve perf here. what's faster? a loop over recursion.

we could also early-return if the parent is a logical expression

Boshen commented:
### Merge activity

* **Jan 10, 8:11 PM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

camc314 committed:
fix(minifier): rotate associative operators to make it more idempotent (#8424)

another **massive** pain to debug

~~https://github.com/oxc-project/monitor-oxc/actions/runs/12717213600~~

https://github.com/oxc-project/monitor-oxc/actions/runs/12717978983/job/35455601146
